xbmc/guilib/GUIVisualisationControl.cpp

```diff
@@ -98,35 +151,324 @@ void CGUIVisualisationControl::Process(unsigned int currentTime, CDirtyRegionLis
+        m_instance->Stop();
+        m_alreadyStarted = false;
       }
 
-      m_bAttemptedLoad = true;
+      std::string strFile = URIUtils::GetFileName(g_application.CurrentFile());
+      m_alreadyStarted = m_instance->Start(m_channels, m_samplesPerSec, m_bitsPerSample, strFile);
+      g_graphicsContext.ApplyStateBlock();
+      m_callStart = false;
+      m_updateTrack = true;
     }
+    else if (m_updateTrack)
+    {
+      /* Initial update of currently processed track */
+      UpdateTrack();
+      m_updateTrack = false;
+    }
+
+    MarkDirtyRegion();
+  }
+
+  CGUIControl::Process(currentTime, dirtyregions);
+}
+
+bool CGUIVisualisationControl::IsDirty()
+{
+  if (m_instance)
+    return m_instance->IsDirty();
+  return false;
+}
+
+void CGUIVisualisationControl::Render()
+{
+  if (m_instance && m_alreadyStarted)
+  {
+    /*
+     * set the viewport - note: We currently don't have any control over how
+     * the addon renders, so the best we can do is attempt to define
+     * a viewport??
+     */
+    g_graphicsContext.SetViewPort(m_posX, m_posY, m_width, m_height);
+    g_graphicsContext.CaptureStateBlock();
+    m_instance->Render();
+    g_graphicsContext.ApplyStateBlock();
+    g_graphicsContext.RestoreViewPort();
   }
-  CGUIRenderingControl::Process(currentTime, dirtyregions);
+  
+  CGUIControl::Render();
+}
+
+void CGUIVisualisationControl::UpdateVisibility(const CGUIListItem *item/* = nullptr*/)
+{
+  // if made invisible, start timer, only free addonptr after
+  // some period, configurable by window class
+  CGUIControl::UpdateVisibility(item);
+  if (!IsVisible() && m_attemptedLoad)
+    FreeResources();
+}
+
+bool CGUIVisualisationControl::CanFocusFromPoint(const CPoint &point) const
+{ // mouse is allowed to focus this control, but it doesn't actually receive focus
+  return IsVisible() && HitTest(point);
 }
 
 void CGUIVisualisationControl::FreeResources(bool immediately)
 {
-  m_bAttemptedLoad = false;
-  // tell our app that we're going
-  if (!m_addon)
+  DeInitVisualization();
+  
+  CGUIControl::FreeResources(immediately);
+
+  CLog::Log(LOGDEBUG, "FreeVisualisation() done");
+}
+
+void CGUIVisualisationControl::OnInitialize(int channels, int samplesPerSec, int bitsPerSample)
+{
+  m_channels = channels;
+  m_samplesPerSec = samplesPerSec;
+  m_bitsPerSample = bitsPerSample;
+  m_callStart = true;
+}
+
+void CGUIVisualisationControl::OnAudioData(const float* pAudioData, int iAudioDataLength)
+{
+  if (!m_instance || !m_alreadyStarted)
+    return;
+
+  // FIXME: iAudioDataLength should never be less than 0
+  if (iAudioDataLength < 0)
+    return;
+
+  // Save our audio data in the buffers
+  std::unique_ptr<CAudioBuffer> pBuffer (new CAudioBuffer(iAudioDataLength));
+  pBuffer->Set(pAudioData, iAudioDataLength);
+  m_vecBuffers.push_back(pBuffer.release());
+
+  if (static_cast<int>(m_vecBuffers.size()) < m_numBuffers)
+    return;
+
+  std::unique_ptr<CAudioBuffer> ptrAudioBuffer(m_vecBuffers.front());
```

Comment:
`std::unique_ptr<CAudioBuffer> ptrAudioBuffer = std::move(m_vecBuffers.front());` after converting to a vector of unique_ptr